File: hackerrank/matrix/diagdiff1.go


// Solution to https://www.hackerrank.com/challenges/diagonal-difference/problem

package main

import (
    "bufio"
    "fmt"
    "io"
    "os"
    "strconv"
    "strings"
    "math"
)

func diagonalDifference(arr [][]int32) int32 {
    fmt.Println("Array Length=", len(arr))
    var d1 int32 = 0
    var d2 int32 = 0
    
    for i:=0; i<len(arr); i++ {
        d1 = d1 + arr[i][i]
        d2 = d2 + arr[i][len(arr)-1-i]
    }
    return int32(math.Abs(float64(d1-d2)))
}

func main() {
    reader := bufio.NewReaderSize(os.Stdin, 1024 * 1024)

    stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
    checkError(err)

    defer stdout.Close()

    writer := bufio.NewWriterSize(stdout, 1024 * 1024)

    nTemp, err := strconv.ParseInt(readLine(reader), 10, 64)
    checkError(err)
    n := int32(nTemp)

    var arr [][]int32
    for i := 0; i < int(n); i++ {
        arrRowTemp := strings.Split(readLine(reader), " ")

        var arrRow []int32
        for _, arrRowItem := range arrRowTemp {
            arrItemTemp, err := strconv.ParseInt(arrRowItem, 10, 64)
            checkError(err)
            arrItem := int32(arrItemTemp)
            arrRow = append(arrRow, arrItem)
        }

        if len(arrRow) != int(n) {
            panic("Bad input")
        }

        arr = append(arr, arrRow)
    }

    result := diagonalDifference(arr)

    fmt.Fprintf(writer, "%d\n", result)

    writer.Flush()
}

func readLine(reader *bufio.Reader) string {
    str, _, err := reader.ReadLine()
    if err == io.EOF {
        return ""
    }

    return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
    if err != nil {
        panic(err)
    }
}

File: hackerrank/matrix/diagdiff1.cpp
/**
  * @brief Solution to https://www.hackerrank.com/challenges/diagonal-difference
  */

#include <bits/stdc++.h>
#include <cmath>

using namespace std;

/**
  * @brief An O(N) Solution which iterates along both the diagonals at the same time 
  */
int diagonalDifference(vector<vector<int>> arr) 
{
    int d1=0; 
    int d2=0; 
    for(int i=0; i<arr.size(); ++i)
    {
        d1 += arr[i][i]; 
        d2 += arr[i][arr.size()-1-i];
    }
    return abs(d1-d2); 
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    int n;
    cin >> n;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    vector<vector<int>> arr(n);
    for (int i = 0; i < n; i++) {
        arr[i].resize(n);

        for (int j = 0; j < n; j++) {
            cin >> arr[i][j];
        }

        cin.ignore(numeric_limits<streamsize>::max(), '\n');
    }

    int result = diagonalDifference(arr);

    fout << result << "\n";

    fout.close();

    return 0;
}
